File: src/_cpp/TowerDefenseBase/scenes/gamegs/MenuTower.cpp
//
//  IslandDefense
//
//  Created by Vladimir Tolmachev on 27.09.14.
//
//
#include "ml/Language.h"
#include "MenuTower.h"
#include "ScoreCounter.h"
#include "GameGS.h"
#include "Tower.h"
#include "support.h"
#include "Tutorial.h"
#include "configuration.h"

NS_CC_BEGIN



MenuTower::MenuTower()
: _waitSellConfirm( false )
, _waitUpgradeConfirm( false )
, _disabled ( false )
{}

MenuTower::~MenuTower()
{
	ScoreCounter::shared().observer( kScoreLevel ).remove( _ID );
}

bool MenuTower::init()
{
	do
	{
		ScrollMenu::init();
		NodeExt::init();

		NodeExt::load( "ini/gamescene", "menutower.xml" );

		_upgrade = getItemByName( "upgrade" );
		_upgradeUn = getItemByName( "upgrade_un" );
		_confirm = getItemByName( "confirm" );
		_confirmUn = getItemByName( "confirm_un" );
		_sell = getItemByName( "sell" );
		_lock = getItemByName( "lock" );

		_upgrade->setCallback( std::bind( &MenuTower::activateUpgrade, this, std::placeholders::_1, true ) );
		_upgradeUn->setCallback( std::bind( &MenuTower::activateUpgrade, this, std::placeholders::_1, false ) );
		_sell->setCallback( std::bind( &MenuTower::activateSell, this, std::placeholders::_1, false ) );
		_lock->setCallback( std::bind( &MenuTower::lockClick, this, std::placeholders::_1 ) );

		_desc.node = getChildByName( "desc" );
		if( _desc.node )
		{
			_desc.name.reset( _desc.node->getChildByName<Label*>( "name" ) );
			_desc.text.reset( _desc.node->getChildByName<Label*>( "text" ) );
			_desc.dmg.reset( _desc.node->getChildByName<Label*>( "dmg" ) );
			_desc.rng.reset( _desc.node->getChildByName<Label*>( "rng" ) );
			_desc.spd.reset( _desc.node->getChildByName<Label*>( "spd" ) );
		}

		setVisible( false );
		return true;
	}
	while( false );
	return false;
}

void MenuTower::appearance()
{
	_disabled = false;
	setEnabled( true );
	setVisible( true );
	hideConfirmButton();
	runEvent( "appearance" );
	ScoreCounter::shared().observer( kScoreLevel ).add( _ID, std::bind( &MenuTower::onChangeMoney, this, std::placeholders::_1 ) );
	onChangeMoney( ScoreCounter::shared().getMoney( kScoreLevel ) );

	_confirm->setVisible( false );
	_confirmUn->setVisible( false );

	scheduleUpdate();
	showRadius( _unit->getPosition(), _unit->getRadius() );
}

void MenuTower::disappearance()
{
	if( _disabled == false )
	{
		setEnabled( false );
		hideConfirmButton();
		runEvent( "disappearance" );
		_unit.reset( nullptr );
		ScoreCounter::shared().observer( kScoreLevel ).remove( _ID );

		_confirm->setVisible( false );
		_confirmUn->setVisible( false );

		_waitSellConfirm = false;
		_waitUpgradeConfirm = false;

		hideRadius();
		unscheduleUpdate();
		_disabled = true;
	}
}

void MenuTower::setUnit( Unit::Pointer unit )
{
	_unit = unit;
	if( _unit )
	{
		std::string cost1 = intToStr( mlTowersInfo::shared().getCost( _unit->getName(), _unit->getLevel() + 1 ) );
		std::string cost2 = intToStr( mlTowersInfo::shared().getSellCost( _unit->getName(), _unit->getLevel() ) );

		_upgrade->getChildByName<Label*>( "cost" )->setString( cost1 );
		_upgradeUn->getChildByName<Label*>( "cost" )->setString( cost1 );
		_sell->getChildByName<Label*>( "cost" )->setString( cost2 );
	}

	if( _unit )
	{
		setPosition( _unit->getPosition() );

		buildDescription( _unit->getLevel() );
	}
}

void MenuTower::buildDescription( unsigned _level )
{
	std::string name = _unit->getName();
	std::string localization = WORD( name + "_name" );
	unsigned level = std::min( _level, _unit->getMaxLevel() );

	std::string dmg = intToStr( mlTowersInfo::shared().get_dmg( name, level ) );
	std::string rng = intToStr( mlTowersInfo::shared().get_rng( name, level ) );
	std::string spd = intToStr( mlTowersInfo::shared().get_spd( name, level ) );
	std::string txt = mlTowersInfo::shared().get_desc( _unit->getName(), 1 );
	if( _desc.name )_desc.name->setString( localization );
	if( _desc.dmg )_desc.dmg->setString( dmg );
	if( _desc.rng )_desc.rng->setString( rng );
	if( _desc.spd )_desc.spd->setString( spd );
	if( _desc.text )_desc.text->setString( txt );
}

void MenuTower::activateUpgrade( Ref * sender, bool availebledButton )
{
	buildDescription( _unit->getLevel() + 1 );
	_confirm->setCallback( std::bind( &MenuTower::confirmUpgrade, this, std::placeholders::_1, true ) );
	_confirmUn->setCallback( std::bind( &MenuTower::confirmUpgrade, this, std::placeholders::_1, false ) );

	_confirm->setVisible( true );
	_confirm->setPosition( _upgrade->getPosition() );
	_confirmUn->setPosition( _upgrade->getPosition() );

	_waitSellConfirm = false;
	_waitUpgradeConfirm = true;
	onChangeMoney( ScoreCounter::shared().getMoney( kScoreLevel ) );

	//float rate = mlTowersInfo::shared().rate( _unit->getName() );
	float radius = mlTowersInfo::shared().radiusInPixels( _unit->getName(), _unit->getLevel() + 1 );
	showRadius( _unit->getPosition( ), radius /** rate*/ );
	runEvent( "onclick" );
}

void MenuTower::confirmUpgrade( Ref * sender, bool availebledButton )
{
	assert( _confirmCurrent );
	assert( _unit );
	auto& board = GameGS::getInstance()->getGameBoard();
	board.upgradeTower( _unit );
	_confirm->setVisible( false );
	_confirmUn->setVisible( false );

	disappearance();

	if( !availebledButton )
	{
        if (k::configuration::useInapps) {
            TutorialManager::shared( ).dispatch( "level_haventgear_upgrade" );
        }
	}
}

void MenuTower::activateSell( Ref * sender, bool availebledButton )
{
	_confirm->setCallback( std::bind( &MenuTower::confirmSell, this, std::placeholders::_1, true ) );

	_confirmUn->setVisible( false );
	_confirm->setVisible( true );
	_confirm->setPosition( _sell->getPosition() );

	_waitSellConfirm = true;
	_waitUpgradeConfirm = false;

	hideRadius();
}

void MenuTower::confirmSell( Ref * sender, bool availebledButton )
{
	assert( _unit );
	GameGS::getInstance()->getGameBoard().removeTower( _unit );
	_confirm->setVisible( false );

	disappearance();
}

void MenuTower::lockClick( Ref * sender )
{
	disappearance();
}

void MenuTower::showConfirmButton()
{}

void MenuTower::hideConfirmButton()
{}

void MenuTower::onChangeMoney( int money )
{
	int cost = mlTowersInfo::shared().getCost( _unit->getName(), _unit->getLevel() + 1 );

	_upgrade->setVisible( cost <= money );
	_upgradeUn->setVisible( cost > money );

	if( !_waitSellConfirm )
	{
		auto hist = _confirmCurrent;
		if( cost <= money )
		{
			_confirmCurrent = _confirm;
			if( _waitUpgradeConfirm )
			{
				_confirm->setVisible( true );
				_confirmUn->setVisible( false );
			}
		}
		else
		{
			_confirmCurrent = _confirmUn;
			if( _waitUpgradeConfirm )
			{
				_confirm->setVisible( false );
				_confirmUn->setVisible( true );
			}
		}
	}

	checkLockedUpgrade();
}

void MenuTower::checkLockedUpgrade()
{
	unsigned level = _unit->getLevel();
	unsigned max = _unit->getMaxLevel();
	unsigned max2 = _unit->getMaxLevelForLevel();

	if( level == max )
	{
		_lock->setVisible( false );
		_upgrade->setVisible( false );
		_upgradeUn->setVisible( false );
	}
	else if( level == max2 )
	{
		_lock->setVisible( true );
		_upgrade->setVisible( false );
		_upgradeUn->setVisible( false );
	}
	else
	{
		_lock->setVisible( false );
	}

}

void MenuTower::update( float )
{
	if( !_unit )
		return;
	if( radiusTowerIsHiden() )
		showRadius( _unit->getPosition(), _unit->getRadius() );

	auto pos = _unit->getPosition();
	auto point = GameGS::getInstance( )->getMainLayer( )->convertToWorldSpace( pos );
	setPosition( point );
	
	if( _desc.node )
	{
		auto screenpos = getPosition( );
		auto dessize = Director::getInstance( )->getOpenGLView( )->getDesignResolutionSize( );

		auto pos = _desc.node->getPosition( );
		pos.x = fabs( pos.x );
		_desc.node->setAnchorPoint( Point( 0, 0.5f ) );
		if( screenpos.x > dessize.width / 2 )
		{
			pos.x = -pos.x;
			_desc.node->setAnchorPoint( Point( 1, 0.5f ) );
		}
		_desc.node->setPosition( pos );
	}
}

NS_CC_END

File: src/_eu/scenes/gamegs/MenuTower.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolm_vl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/

//Define namespace
var EU = EU || {};




EU.MenuTower = EU.ScrollMenu.extend( // public NodeExt
{
    /** @type {EU.Unit} */ _unit : null,
    /** @type {Boolean} */ _disabled : null,
    /** @type {MenuItem} */ _upgrade : null,
    /** @type {MenuItem} */ _upgradeUn : null,
    /** @type {MenuItem} */ _sell : null,
    /** @type {MenuItem} */ _confirm : null,
    /** @type {MenuItem} */ _confirmUn : null,
    /** @type {MenuItem} */ _confirmCurrent : null,
    /** @type {MenuItem} */ _lock : null,
    /** @type {Boolean} */ _waitSellConfirm : null,
    /** @type {Boolean} */ _waitUpgradeConfirm : null,
    _desc : function() {
        "use strict";

        /**Node*/ this.node = null;
         /**Label*/ this.name = null;
         /**Label*/ this.text = null;
         /**Label*/ this.dmg = null;
         /**Label*/ this.rng = null;
         /**Label*/ this.spd = null;
    },


    onExit: function()
    {
        EU.ScoreCounter.observer( EU.kScoreLevel ).remove( this.__instanceId );
    },

    ctor: function()
    {
        this._waitSellConfirm = false;
        this._waitUpgradeConfirm = false;
        this._disabled = false;

        this._super();

        this.initExt();

        this.load_str_n_str( "ini/gamescene", "menutower.xml" );

        this._upgrade = this.getItemByName( "upgrade" );
        this._upgradeUn = this.getItemByName( "upgrade_un" );
        this._confirm = this.getItemByName( "confirm" );
        this._confirmUn = this.getItemByName( "confirm_un" );
        this._sell = this.getItemByName( "sell" );
        this._lock = this.getItemByName( "lock" );

        var self = this;
        this._upgrade.setCallback( this.activateUpgrade.bind(this, true), this );
        this._upgradeUn.setCallback( this.activateUpgrade.bind(this, false), this );
        this._sell.setCallback( this.activateSell.bind(this, false), this );
        this._lock.setCallback( this.lockClick, this);

        this._desc.node = this.getChildByName( "desc" );
        if( this._desc.node )
        {
            this._desc.name = ( this._desc.node.getChildByName( "name" ) );
            this._desc.text = ( this._desc.node.getChildByName( "text" ) );
            this._desc.dmg = ( this._desc.node.getChildByName( "dmg" ) );
            this._desc.rng = ( this._desc.node.getChildByName( "rng" ) );
            this._desc.spd = ( this._desc.node.getChildByName( "spd" ) );
        }

        this.setVisible( false );
        return true;
    },

    appearance: function()
    {
        this._disabled = false;
        this.setEnabled( true );
        this.setVisible( true );
        this.hideConfirmButton();
        this.runEvent( "appearance" );
        EU.ScoreCounter.observer( EU.kScoreLevel ).add( this.__instanceId, this.onChangeMoney, this );
        this.onChangeMoney( EU.ScoreCounter.getMoney( EU.kScoreLevel ) );

        this._confirm.setVisible( false );
        this._confirmUn.setVisible( false );

        this.scheduleUpdate();
        EU.showRadius( this._unit.getPosition(), this._unit._radius );
    },

    disappearance: function()
    {
        if( this._disabled == false )
        {
            this.setEnabled( false );
            this.hideConfirmButton();
            this.runEvent( "disappearance" );
            this._unit = ( null );
            EU.ScoreCounter.observer( EU.kScoreLevel ).remove( this.__instanceId );

            this._confirm.setVisible( false );
            this._confirmUn.setVisible( false );

            this._waitSellConfirm = false;
            this._waitUpgradeConfirm = false;

            EU.hideRadius();
            this.unscheduleUpdate();
            this._disabled = true;
        }
    },

    setUnit: function( unit )
    {
        this._unit = unit;
        if( this._unit )
        {
            var cost1 = ( EU.mlTowersInfo.getCost( this._unit.getName(), this._unit._level + 1 ) );
            var cost2 = ( EU.mlTowersInfo.getSellCost( this._unit.getName(), this._unit._level ) );

            this._upgrade.getChildByName( "cost" ).setString( cost1 );
            this._upgradeUn.getChildByName( "cost" ).setString( cost1 );
            this._sell.getChildByName( "cost" ).setString( cost2 );
        }

        if( this._unit )
        {
            this.setPosition( this._unit.getPosition() );

            this.buildDescription( this._unit._level );
        }
    },

    buildDescription: function( _level )
    {
        var name = this._unit.getName();
        var localization = EU.Language.string( name + "_name" );
        var level = Math.min( this._level, this._unit._maxLevel );

        var dmg = ( EU.mlTowersInfo.get_dmg( name, level ) );
        var rng = ( EU.mlTowersInfo.get_rng( name, level ) );
        var spd = ( EU.mlTowersInfo.get_spd( name, level ) );
        var txt = EU.mlTowersInfo.get_desc( this._unit.getName(), 1 );
        if( this._desc.name )this._desc.name.setString( localization );
        if( this._desc.dmg )this._desc.dmg.setString( dmg );
        if( this._desc.rng )this._desc.rng.setString( rng );
        if( this._desc.spd )this._desc.spd.setString( spd );
        if( this._desc.text )this._desc.text.setString( txt );
    },

    activateUpgrade: function( availebledButton, sender )
    {
        this.buildDescription( this._unit._level + 1 );
        var self = this;
        this._confirm.setCallback( this.confirmUpgrade.bind(this, true) );
        this._confirmUn.setCallback( this.confirmUpgrade.bind(this, false) );

        this._confirm.setVisible( true );
        this._confirm.setPosition( this._upgrade.getPosition() );
        this._confirmUn.setPosition( this._upgrade.getPosition() );

        this._waitSellConfirm = false;
        this._waitUpgradeConfirm = true;
        this.onChangeMoney( EU.ScoreCounter.getMoney( EU.kScoreLevel ) );

        //var rate = EU.mlTowersInfo.rate( this._unit.getName() );
        var radius = EU.mlTowersInfo.radiusInPixels( this._unit.getName(), this._unit._level + 1 );
        EU.showRadius( this._unit.getPosition( ), radius /** rate*/ );
        this.runEvent( "onclick" );
        sender.setVisible(false);
    },

    confirmUpgrade: function( availebledButton, sender )
    {
        EU.assert( this._confirmCurrent );
        EU.assert( this._unit );
        var board = EU.GameGSInstance.getGameBoard();
        board.upgradeTower( this._unit );
        this._confirm.setVisible( false );
        this._confirmUn.setVisible( false );

        this.disappearance();

        if( !availebledButton )
        {
            if (EU.k.useInapps) {
                EU.TutorialManager.dispatch( "level_haventgear_upgrade" );
            }
        }
    },

    activateSell: function( availebledButton, sender )
    {
        var self = this;
        this._confirm.setCallback( this.confirmSell.bind(this, true) );

        this._confirmUn.setVisible( false );
        this._confirm.setVisible( true );
        this._confirm.setPosition( this._sell.getPosition() );

        this._waitSellConfirm = true;
        this._waitUpgradeConfirm = false;

        EU.hideRadius();
        sender.setVisible(false);
    },

    confirmSell: function( availebledButton, sender )
    {
        EU.assert( this._unit );
        EU.GameGSInstance.getGameBoard().removeTower( this._unit );
        this._confirm.setVisible( false );

        this.disappearance();
    },

    lockClick: function( sender )
    {
        this.disappearance();
    },

    showConfirmButton: function()
    {},

    hideConfirmButton: function()
    {},

    onChangeMoney: function(money )
    {
        var cost = EU.mlTowersInfo.getCost( this._unit.getName(), this._unit._level + 1 );

        this._upgrade.setVisible( cost <= money );
        this._upgradeUn.setVisible( cost > money );

        if( !this._waitSellConfirm )
        {
            var hist = this._confirmCurrent;
            if( cost <= money )
            {
                this._confirmCurrent = this._confirm;
                if( this._waitUpgradeConfirm )
                {
                    this._confirm.setVisible( true );
                    this._confirmUn.setVisible( false );
                }
            }
            else
            {
                this._confirmCurrent = this._confirmUn;
                if( this._waitUpgradeConfirm )
                {
                    this._confirm.setVisible( false );
                    this._confirmUn.setVisible( true );
                }
            }
        }

        this.checkLockedUpgrade();
    },

    checkLockedUpgrade: function()
    {
        var level = this._unit._level;
        var max = this._unit._maxLevel;
        var max2 = this._unit._maxLevelForLevel;

        if( level == max )
        {
            this._lock.setVisible( false );
            this._upgrade.setVisible( false );
            this._upgradeUn.setVisible( false );
        }
        else if( level == max2 )
        {
            this._lock.setVisible( true );
            this._upgrade.setVisible( false );
            this._upgradeUn.setVisible( false );
        }
        else
        {
            this._lock.setVisible( false );
        }

    },

    update: function( )
    {
        if( !this._unit )
            return;
        if( EU.radiusTowerIsHiden() )
            EU.showRadius( this._unit.getPosition(), this._unit._radius );

        var pos = this._unit.getPosition();
        var point = EU.GameGSInstance.getMainLayer( ).convertToWorldSpace( pos );
        this.setPosition( point );

        if( this._desc.node )
        {
            var screenpos = this.getPosition( );
            var dessize = cc.view.getDesignResolutionSize( );

            var pos = this._desc.node.getPosition( );
            pos.x = Math.abs( pos.x );
            this._desc.node.setAnchorPoint( cc.p( 0, 0.5 ) );
            if( screenpos.x > dessize.width / 2 )
            {
                pos.x = -pos.x;
                this._desc.node.setAnchorPoint( cc.p( 1, 0.5 ) );
            }
            this._desc.node.setPosition( pos );
        }
    }
});

EU.NodeExt.call(EU.MenuTower.prototype);
